### ISSUE TITLE
uncertain lint error reporting

### ISSUE THREAD

**Issue by zchcai** [google/fuzzbench#555]
Uncertain behavior for `make lint`.
Sometimes it will give this error, which doesn't make sense as the later execution does call fake_jobs module.
```
source .venv/bin/activate && python3 presubmit.py lint
************* Module run_experiment
fuzzbench/run_experiment.py:21:0: E0611: No name 'fake_jobs' in module 'fuzzbench' (no-name-in-module)
```

Any clues or ideas to fix this?

**Comment by jonathanmetzman** [comment]
You might need to add an `__init__.py` to `fuzzbench`.

**Comment by zchcai** [comment]
An interesting behavior is, google/master works and if I switch from google/master to one pre-failed branch, it works too.

**Comment by zchcai** [comment]
> An interesting behavior is, google/master works and if I switch from google/master to one pre-failed branch, it works too. (Without adding **init**.py)

I think this is just fake success since we don't change any source files. Then the lint doesn't work.

**Comment by zchcai** [comment]
As suggested offline, we may need to change the folder name for the inside `fuzzbench`. See #559. This error is related to the name.

**Comment by zchcai** [comment]
As suggested offline, we may want to try the latest pylint 2.5.3, as we're using 2.4.4 now.

**Comment by jonathanmetzman** [comment]
Feel free to update it in requirements.txt

**Comment by zchcai** [comment]
> As suggested offline, we may want to try the latest pylint 2.5.3, as we're using 2.4.4 now.

This probably not the reason.
When I change it, there is additional errors:
```
~/fuzzbench$ make lint
source .venv/bin/activate && python3 presubmit.py lint
************* Module run_experiment
fuzzbench/run_experiment.py:23:0: E0401: Unable to import 'common' (import-error)
fuzzbench/run_experiment.py:24:0: E0611: No name 'fake_jobs' in module 'fuzzbench' (no-name-in-module)

------------------------------------------------------------------
Your code has been rated at 7.78/10 (previous run: 7.78/10, +0.00)

ERROR: lint failed, see errors above.
Failed.
make: *** [Makefile:60: lint] Error 1
```
Any thoughts?

**Comment by lszekeres** [comment]
> > As suggested offline, we may want to try the latest pylint 2.5.3, as we're using 2.4.4 now.
> 
> This probably not the reason.
> When I change it, there is additional errors:
> 
> ```
> ~/fuzzbench$ make lint
> source .venv/bin/activate && python3 presubmit.py lint
> ************* Module run_experiment
> fuzzbench/run_experiment.py:23:0: E0401: Unable to import 'common' (import-error)
> fuzzbench/run_experiment.py:24:0: E0611: No name 'fake_jobs' in module 'fuzzbench' (no-name-in-module)
> 
> ------------------------------------------------------------------
> Your code has been rated at 7.78/10 (previous run: 7.78/10, +0.00)
> 
> ERROR: lint failed, see errors above.
> Failed.
> make: *** [Makefile:60: lint] Error 1
> ```
> 
> Any thoughts?

it "works on my machine". what's your python version? 
mine is
```
$ source .venv/bin/activate && python3 --version
Python 3.7.7
```

**Comment by zchcai** [comment]
Mine is 3.8.4rc1. Interesting! Mine should have be 3.7.x too.
```
$ source .venv/bin/activate && python3 --version
Python 3.8.4rc1

**Comment by zchcai** [comment]
> it "works on my machine". 

Could you tell me your steps to check it works? If there is no change in fuzzbench/run_experiment.py, then it will not report that error.

**Comment by lszekeres** [comment]
> > it "works on my machine".
> 
> Could you tell me your steps to check it works? If there is no change in fuzzbench/run_experiment.py, then it will not report that error.

```
source .venv/bin/activate && python3 -m pylint fuzzbench/run_experiment.py
```

**Comment by zchcai** [comment]
From offline discussion:
Use this [code repo](https://github.com/zchcai/fuzzbench/tree/36f6c6c2e6636cc5199d1b7a6374d679ef06c8fb) and change a bit in `fuzzbench/run_experiment.py`.

1. make lint fail
```
~/fuzzbench$ make lint
source .venv/bin/activate && python3 presubmit.py lint
************* Module run_experiment
fuzzbench/run_experiment.py:22:0: E0611: No name 'fake_jobs' in module 'fuzzbench' (no-name-in-module)

-------------------------------------------------------------------
Your code has been rated at 8.89/10 (previous run: 10.00/10, -1.11)

ERROR: lint failed, see errors above.
Failed.
make: *** [Makefile:60: lint] Error 1
```

2. single work
```
~/fuzzbench$ . .venv/bin/activate && python3 -m pylint fuzzbench/run_experiment

-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 8.89/10, +1.11)

~/fuzzbench$ . .venv/bin/activate && python3 -m pylint common/config_utils.py

------------------------------------
Your code has been rated at 10.00/10
```

3. the simplest fail case
```
~/fuzzbench$ . .venv/bin/activate && python3 -m pylint common/config_utils.py fuzzbench/run_experiment.py
************* Module run_experiment
fuzzbench/run_experiment.py:22:0: E0611: No name 'fake_jobs' in module 'fuzzbench' (no-name-in-module)

-------------------------------------------------------------------
Your code has been rated at 8.53/10 (previous run: 10.00/10, -1.47)
```

So we may need to pylint one file each time, not together.

**Comment by lszekeres** [comment]
TL;DR: delete `common/__init__.py` and it should work.

Currently `presubmit.py` calls `pylint` by passing individual .py paths. This is not fully supported as it relies on `pylint` guessing properly the package/module boundaries based on the passed paths: https://docs.pylint.org/en/1.6.0/faq.html#can-i-give-pylint-a-file-as-an-argument-instead-of-a-module. That `__init__.py` file in `common` confuses pylint about what's a package and what's not. To fix this we can just remove `__init__.py` files for now. Later, once we've migrated all python code under the new `fuzzbench` sub-directory, that will be our "python package", and we can call pylint properly, by simply passing the package name to it.

**Comment by zchcai** [comment]
> TL;DR: delete common/__init__.py and it should work.

In my machine, I just check with [this code repo](https://github.com/zchcai/fuzzbench/tree/36f6c6c2e6636cc5199d1b7a6374d679ef06c8fb) and it will fail more when deleting `__init__.py` under `common` directory.
```
$ rm common/__init__.py
$ make lint
source .venv/bin/activate && python3 presubmit.py lint
************* Module run_experiment
fuzzbench/run_experiment.py:21:0: E0611: No name 'config_utils' in module 'common' (no-name-in-module)
fuzzbench/run_experiment.py:21:0: E0611: No name 'environment' in module 'common' (no-name-in-module)
fuzzbench/run_experiment.py:21:0: E0611: No name 'yaml_utils' in module 'common' (no-name-in-module)

------------------------------------------------------------------
Your code has been rated at 6.67/10 (previous run: 8.89/10, -2.22)

ERROR: lint failed, see errors above.
Failed.
make: *** [Makefile:60: lint] Error 1
```

**Comment by lszekeres** [comment]
> $ rm common/__init__.py

to fix `make lint` fully (not just `python3 -m pylint common/config_utils.py fuzzbench/run_experiment.py`) you need to remove all `__init__.py` files. can you try deleting `experiment/__init__.py` too and try again?

> In my machine, I just check with [this code repo](https://github.com/zchcai/fuzzbench/tree/36f6c6c2e6636cc5199d1b7a6374d679ef06c8fb) and it will fail more when deleting 

if the above doesn't work, let me know which branch is that. i tested this in your `default-e2e-test-config` branch before

**Comment by jonathanmetzman** [comment]
Could we leave the __init__.py files and simply pass (the outer) fuzzbench as the package?
I mentioned elsewhere (https://github.com/google/fuzzbench/issues/368) how linting only changed files is a bug because when you change a file you can break dependent modules and these won't be linted unless changed.

**Comment by zchcai** [comment]
After removing existing `__init__.py` and ensure the Python version as 3.7.7. Then this bug should be resolved.

**Comment by lszekeres** [comment]
Here are a couple things to consider:

- We should either use init files everywhere or rely on implicit namespaces (no init files) everywhere. Mixing them will lead to unexpected results with pylint (see next points).

- The problem with init files is that if we forget to put it in some directory (or subdirectory) pylint will just silently skip linting those files. We might not notice these for a long time. For example it looks like we've been missing the `fuzzers/` directory all this time: https://github.com/google/fuzzbench/pull/566)

- The benefit of not having any init files and relying on implicit namespaces is that we don't need to remember adding an init file to every single new directory: if we run `pylint <package>` everything will be recursively checked under `<package>`

- The problem with doing `pylint ../fuzzbench-repo` is that we don't control the name 'fuzzbench-repo'. One might checkout the repo to whatever directory.

- My understanding of the current issue with pylint is that by default the repo name is fuzzbench, and then we have a fuzzbench directory in the repo. Now, when we rely on the (semi-supported) feature of passing in individual file paths, pylint's heuristic will sometimes decide that the "fuzzbench" package is under `fuzzbench/fuzzbench`, and sometimes it'll decide it's under `fuzzbench` (the repo). I'm guessing this conflict causes the non-deterministic behavior. 

- Probably the most ideal would be if we used the convention of having a top level python package in the repo, ie. a "fuzzbench" directory, so we could just `pylint fuzzbench`. I created an issue to track this: https://github.com/google/fuzzbench/issues/574

- Before we get there though, perhaps the best thing we can do is to call pylint by listing (sub-)packages, ie. `pylint analysis common database ...`. I updated  https://github.com/google/fuzzbench/pull/566 with that, PTAL. (It seems to work, but i'm not 100% sure it completely eliminates the non-deterministic issue as i also list 'presubmit.py' from the repo root).
